File: Dynamic-Programming/139-M-Word-Break/word_break_dp.py
class Solution:
    def wordBreak(self, s: str, wordDict: list[str]) -> bool:
        n = len(s)
        if n==0:
            return True
        my_set = set(wordDict)
        OPT = [False] * (n+1)
        OPT[0] = True
        
        for i in range(1, n+1):
            for j in range(i):
                if OPT[j] == True and s[i:j] in my_set:
                    OPT[i] = True
                    break
        return OPT[n]

File: Dynamic-Programming/139-M-Word-Break/word_break_dp.cpp
class Solution {
public:
    bool wordBreak(string s, vector<string>& wordDict) {
        int n = s.length();
        if (n==0) return true;
        unordered_set<string> my_set(wordDict.begin(), wordDict.end());
        vector<bool> OPT(n+1, false);
        OPT[0] = true;

        for (int i=1; i<=n; i++) {
            for (int j=0; j<i; j++) {
                if (OPT[j]==true && my_set.count(s.substr(j, i-j))) {
                    OPT[i] = true;
                    break;
                }
            }
        }
        return OPT[n];
    }
};